### PR TITLE
Appendix F: Fuzzing a Palindrome Integer checker

### PR DESCRIPTION
**Pull request by doniacld**
### Requirement
- Check if the string input is a palindrome integer
- `01` is not a palindrome
- `010` is a palindrome

### Steps

1. Basic flip working with integesr
2. Convert to runes and compare the first and last digits both ways

### CODE REVIEW

File: appendices/F/4_fix_sign/palindrome/palindrome_test.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+	f.Add("01")   // pads with 0
+	f.Add("\xff") // not a number
+	f.Add("+1")   // signed +
+	f.Add("-1")   // signed -
+
+	// Call Fuzz and verify output
+	f.Fuzz(func(t *testing.T, input string) {
+		got, err := IsPalindromeNumber(input)
+		if err != nil {
+			// V2 ignore if it's not a number
```
Comment (doniacld) [inline appendices/F/4_fix_sign/palindrome/palindrome_test.go:22]: To clean

File: appendices/F/5_fix_too_big/palindrome/palindrome_test.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
+package palindrome
+
+import (
+	"slices"
+	"testing"
+)
+
+func FuzzIsPalindromeNumber(f *testing.F) {
+	// Seed corpus with some basic test cases
```
Comment (doniacld) [inline appendices/F/5_fix_too_big/palindrome/palindrome_test.go:9]: Remove duplicate

File: appendices/F/1_flip_int/go.mod [hunk 3]
```diff
@@ -0,0 +1,3 @@
+module github.com/ablqk/tiny-go-projects/appendices/F
+
+go 1.22.4
```
Comment (floppyzedolfin) [inline appendices/F/1_flip_int/go.mod:3]:
```suggestion
go 1.23.0
```
(and subsequent files)

File: appendices/F/1_flip_int/palindrome/palindrome.go [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+	if err != nil {
+		return false
+	}
+
+	// flip the integer and compare
+	original := toInt
+	flip := 0
+	for toInt > 0 {
+		flip = 10*flip + toInt%10 // 0*10 + 1234%10 = 4
+		toInt = toInt / 10        // 1234/10 = 123
```
Comment (floppyzedolfin) [inline appendices/F/1_flip_int/palindrome/palindrome.go:21]:
```suggestion
		flip = 10*flip + toInt%10 //   4 | 43 | 432 | 4321
		toInt = toInt / 10        // 123 | 12 |   1 |    0
```

File: appendices/F/1_flip_int/palindrome/testdata/fuzz/FuzzIsPalindromeNumber/7f5131ea0ab9779f [hunk 5]
```diff
@@ -0,0 +1,2 @@
+go test fuzz v1
```
Comment (floppyzedolfin) [inline appendices/F/1_flip_int/palindrome/testdata/fuzz/FuzzIsPalindromeNumber/7f5131ea0ab9779f:1]: do we want to commit this file?

File: appendices/F/1_flip_int/palindrome/palindrome_test.go [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+import (
+	"slices"
+	"testing"
+)
+
+func FuzzIsPalindromeNumber(f *testing.F) {
+	// Seed corpus with some basic test cases
+	f.Add("1221") // nominal case
+	f.Add("")     // empty string
```
Comment (floppyzedolfin) [inline appendices/F/1_flip_int/palindrome/palindrome_test.go:11]: We add this test case here as an edge case that checks that our code is "soild" enough.

File: appendices/F/1_flip_int/palindrome/palindrome_test.go [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+	// Call Fuzz and verify output
```
Comment (floppyzedolfin) [inline appendices/F/1_flip_int/palindrome/palindrome_test.go:13]:
```suggestion
```

File: appendices/F/3_fix_string/palindrome/palindrome.go [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+// "-1" is not a palindrome
+func IsPalindromeNumber(s string) (bool, error) {
+	// check if it's a number, e.g.: "ң\xa3" is not a number
+	_, err := strconv.Atoi(s)
+	if err != nil {
+		// returns an error if it's not a number
+		return false, fmt.Errorf("conversion failed %s: %w", s, err)
+	}
+
+	for i := 0; i < len(s)/2; i++ {
```
Comment (floppyzedolfin) [inline appendices/F/3_fix_string/palindrome/palindrome.go:21]: I think this should iterate over `[]rune(s)`, not over `s`. The impact is minimal for our Pali function, but it's slightly better with `[]rune`:


```go
runes := []rune(s)
length := len(runes)
for idx := 0; idx < length / 2; idx++ {
  if runes[idx] != runes[length - 1 - idx] {
    return false, nil
  }
}
```

File: appendices/F/5_fix_too_big/palindrome/palindrome.go [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+)
+
+// IsPalindromeNumber returns if an integer is readable in both ways.
+// Only unsigned int are supported.
+// "1221" is a palindrome
+// "01" is not a palindrome
+// "-1" is not a palindrome
+func IsPalindromeNumber(s string) (bool, error) {
+	// value could be bigger than max int e.g.: 10000000000000000000
+	if len(s) > 10 {
```
Comment (floppyzedolfin) [inline appendices/F/5_fix_too_big/palindrome/palindrome.go:16]: (we should still hit some errors here - `2^32+1` is ~ `4 * 10 ^9`. But it's a safe test.

**Review by floppyzedolfin — APPROVED** [review]

File: appendices/F/1_flip_int/palindrome/palindrome_test.go [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+	f.Fuzz(func(t *testing.T, input string) {
+		got := IsPalindromeNumber(input)
+
+		// Basic validation: a string reversed should match the palindrome status
+		reversed := reverseInt(input)
+		want := IsPalindromeNumber(reversed)
```
Comment (doniacld) [inline appendices/F/1_flip_int/palindrome/palindrome_test.go:19]: Update naming

File: appendices/F/1_flip_int/palindrome/palindrome_test.go [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+		if got != want {
+			t.Errorf("Palindrome mismatch for input: %s and its reverse: %s", input, reversed)
+		}
+	})
+}
+
+// reverseInt reverses an integer using the built-in slices.Reverse function
+func reverseInt(s string) string {
+	split := []rune(s)
```
Comment (doniacld) [inline appendices/F/1_flip_int/palindrome/palindrome_test.go:28]: runes

File: appendices/F/3_fix_string/palindrome/palindrome_test.go [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+	f.Add("")     // empty string
+	f.Add("10")   // ends with 0
+	f.Add("01")   // pads with 0
+	f.Add("\xff") // not a number
+
+	// Call Fuzz and verify output
+	f.Fuzz(func(t *testing.T, input string) {
+		got, err := IsPalindromeNumber(input)
+		if err != nil {
+			// ignore if it's not a number
```
Comment (doniacld) [inline appendices/F/3_fix_string/palindrome/palindrome_test.go:20]: update the comment `Ignore if the input is not a number`

File: appendices/F/3_fix_string/palindrome/palindrome.go [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+import (
+	"fmt"
+	"strconv"
+)
+
+// IsPalindromeNumber returns if an integer is readable in both ways.
+// Only unsigned int are supported.
+// "1221" is a palindrome
+// "01" is not a palindrome
+// "-1" is not a palindrome
```
Comment (doniacld) [inline appendices/F/3_fix_string/palindrome/palindrome.go:12]: Too early?

File: appendices/F/1_flip_int/palindrome/palindrome_test.go [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+		if got != reverseIsPalindrome {
+			t.Errorf("Palindrome mismatch for input: "+
+				"%s (isPalindrome: %v) and "+
+				"its reverse: %s (isPalindrome: %v)",
+				input, got, reversed, reverseIsPalindrome)
+		}
+	})
+}
+
+// reverse reverses a string using the built-in slices.Reverse function
```
Comment (floppyzedolfin) [inline appendices/F/1_flip_int/palindrome/palindrome_test.go:28]:
```suggestion
// reverse reverses a string using the standard library slices.Reverse function
```

It's not a buil-tin

File: appendices/F/1_flip_int/palindrome/palindrome_test.go [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+	f.Fuzz(func(t *testing.T, input string) {
+		got := IsPalindromeNumber(input)
+
+		// Basic validation: a string reversed should match the palindrome status
+		reversed := reverse(input)
+		reverseIsPalindrome := IsPalindromeNumber(reversed)
+		if got != reverseIsPalindrome {
+			t.Errorf("Palindrome mismatch for input: "+
+				"%s (isPalindrome: %v) and "+
+				"its reverse: %s (isPalindrome: %v)",
```
Comment (floppyzedolfin) [inline appendices/F/1_flip_int/palindrome/palindrome_test.go:22]:
```suggestion
				"%s (isPalindrome: %t) and "+
				"its reverse: %s (isPalindrome: %t)",
```

File: appendices/F/2_0_fix_pad/palindrome/palindrome_test.go [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+	f.Add("1_221")               // a valid int with "_"
+
+	f.Fuzz(func(t *testing.T, input string) {
+		got := IsPalindromeNumber(input)
+
+		// Basic validation: a string reversed should match the palindrome status
+		reversed := reverse(input)
+		reverseIsPalindrome := IsPalindromeNumber(reversed)
+		if got != reverseIsPalindrome {
+			t.Errorf("Palindrome mismatch for input: %s (isPalindrome: %v) and its reverse: %s (isPalindrome: %v)", input, got, reversed, reverseIsPalindrome)
```
Comment (floppyzedolfin) [inline appendices/F/2_0_fix_pad/palindrome/palindrome_test.go:34]:
```suggestion
			t.Errorf("Palindrome mismatch for input: %s (isPalindrome: %t) and its reverse: %s (isPalindrome: %t)", input, got, reversed, reverseIsPalindrome)
```

**Review by floppyzedolfin — APPROVED** [review]
Seems good :)
(although I'd call it "trailing zeroes" rather than padding -- padding is when we know we want a specific size - in the case raised by fuzzing, the intention of having a fixed length isn't really obvious)

File: appendices/F/1_flip_int/palindrome/testdata/fuzz/FuzzIsPalindromeNumber/7f5131ea0ab9779f (hunk 5 above)
Comment (doniacld) [reply appendices/F/1_flip_int/palindrome/testdata/fuzz/FuzzIsPalindromeNumber/7f5131ea0ab9779f:1]: For me yes.

File: appendices/F/1_flip_int/palindrome/palindrome.go [hunk 17]
```diff
@@ -0,0 +1,25 @@
+package palindrome
+
+import (
+	"strconv"
+)
+
+// IsPalindromeNumber returns if an integer is readable in both ways.
```
Comment (alienorlatour) [inline appendices/F/1_flip_int/palindrome/palindrome.go:7]:
```suggestion
// IsPalindromeNumber returns whether an integer is identical when read both ways.
```

File: appendices/F/1_flip_int/palindrome/testdata/fuzz/FuzzIsPalindromeNumber/7f5131ea0ab9779f (hunk 5 above)
Comment (alienorlatour) [reply appendices/F/1_flip_int/palindrome/testdata/fuzz/FuzzIsPalindromeNumber/7f5131ea0ab9779f:1]: What does it do?

File: appendices/F/2_0_fix_trailing_zeroes/go.mod [hunk 18]
```diff
@@ -0,0 +1,3 @@
+module learngo-pockets
```
Comment (alienorlatour) [inline appendices/F/2_0_fix_trailing_zeroes/go.mod:1]:
```suggestion
module learngo-pockets/f-fuzzing
```

File: appendices/F/1_flip_int/palindrome/testdata/fuzz/FuzzIsPalindromeNumber/7f5131ea0ab9779f (hunk 5 above)
Comment (doniacld) [reply appendices/F/1_flip_int/palindrome/testdata/fuzz/FuzzIsPalindromeNumber/7f5131ea0ab9779f:1]: This is the generated file by the go test -fuzz command with the failing test.
